java/test/com/facebook/watchman/WatchmanConnectionTest.java

```diff
@@ -295,5 +297,83 @@ public void onFailure(Throwable t) {
       entry++;
     }
   }
+
+  @Test
+  public void propagatesDeserializationException() throws Exception {
+    final AtomicReference<BserDeserializer.BserEofException> expectedExceptionRef = new AtomicReference<>();
+    final Semaphore commandSentSemaphore = new Semaphore(0);
+    WatchmanConnection.WatchmanCommandListener commandListener =
+        new WatchmanConnection.WatchmanCommandListener() {
+          @Override public void onStart() { }
+          @Override public void onSent() { commandSentSemaphore.release(); }
+          @Override public void onReceived() { }
+        };
+    Callable<Map<String, Object>> incomingMessageGetter = new Callable<Map<String, Object>>() {
+      @Override
+      public Map<String, Object> call() throws Exception {
+        commandSentSemaphore.acquire();
+        BserDeserializer.BserEofException e = new BserDeserializer.BserEofException("expected");
+        expectedExceptionRef.set(e);
+        throw e;
+      }
+    };
+    WatchmanConnection connection = new WatchmanConnection(
+        incomingMessageGetter,
+        mOutgoingMessageStream,
+        Optional.<WatchmanConnection.WatchmanCommandListener>of(commandListener));
+    connection.start();
+    ListenableFuture<Map<String, Object>> listenableFuture = connection.run(Arrays.asList("clock", "/a/b/c"));
+    try {
+      Map<String, Object> receivedResponse = listenableFuture.get(5, TimeUnit.SECONDS);
+      Assert.fail("Should not get a response, got " + receivedResponse.toString());
+    } catch (ExecutionException executionException) {
+      Assert.assertSame(expectedExceptionRef.get(), executionException.getCause());
+    }
+  }
+
+  @Test
+  public void propagatesInterruptionException() throws Exception {
+    final Semaphore enteredCallable = new Semaphore(0);
+    final Semaphore exitedCallable = new Semaphore(0);
+    final AtomicReference<Thread> threadRef = new AtomicReference<>();
+    final AtomicReference<InterruptedException> expectedException = new AtomicReference<>();
+    Callable<Map<String, Object>> incomingMessageGetter = new Callable<Map<String, Object>>() {
+      @Override
+      public Map<String, Object> call() throws Exception {
+        try {
+          threadRef.set(Thread.currentThread());
+          enteredCallable.release();
+          Thread.sleep(1);
```

Comment:
Can we avoid sleeping in a test? Maybe use a latch to make sure thread is at a specific location and then interrupt it? Sleeping is usually associated with non-determinism and flakiness in tests.